Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` an `(S,R)`-bimodule.
+Consider the tensor functor `(X ⊗[R] ·)` from the category of left `R`-module to the category of
+left `S`-module and the hom functor `(X →ₗ[S] ·)` from the category of left `S`-module to the
+category of left `R`-module. They form an adjunction. In particular we have that
+```
+Hom_S(X⊗[R]Y, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+
+## Implementation notes
+
+1. Order of arguments
+```
+Hom_S(Y⊗[R]X, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+is perhaps more natural to work with because all the argument have the same order.
+But currently mathlib4 does not know `Y⊗[R] X` is an right `S`-module as well.
+
+2. Why not use the [`Tower` file](Mathlib/LinearAlgebra/TensorProduct/Tower.lean`)
+
+In our setting `X` is an `(R, S)`-bimodule and `Y` an `R`-module and `Z` an `S`-module
+so to use the `Tower` file, we need `S` to be an `R`-algebra which is a luxury we do not have.
+But note that in `Tower` file, `curry` and `uncurry` are both tri-linear maps. So `Tower` file
+allows interplay of 3 rings which is not allowed in this file.
+
+-/
+
+universe u u' v v' v''
+
+open TensorProduct CategoryTheory
+
+variable (R : Type u) (S : Type u') (X : Type v)
+variable [CommRing R] [Ring S]
+variable [AddCommGroup X] [Module Rᵐᵒᵖ X] [Module S X] [SMulCommClass S Rᵐᵒᵖ X]
+
+open MulOpposite
+
+-- TODO: remove these local instances when TensorProduct is generalized over non-commutative rings.
+
+/-- The `R`-module structure on a `Rᵐᵒᵖ`-module `X` for a commutative ring `R`. -/
+def Module.ofMop : Module R X :=
+  Module.compHom X ((RingHom.id R).toOpposite fun _ _ => mul_comm _ _)
+
+attribute [local instance] Module.ofMop
+
+variable {Y : Type v'} [AddCommGroup Y] [Module R Y]
+
+/-- A version of `TensorProduct.leftModule` with `Rᵐᵒᵖ` in place of `R`. -/
+noncomputable def Module.tensorLeftOfMop : Module S (X ⊗[R] Y) :=
+  have : SMulCommClass R S X :=
+    ⟨fun r s x ↦ show (op r) • s • x = s • (op r) • x by rw [← smul_comm]⟩
+  inferInstance
+
+attribute [local instance] Module.tensorLeftOfMop
+
+@[simp] lemma tensorModSMul_smul_tmul
+  (s : S) (x : X) (y : Y) : s • (x ⊗ₜ[R] y) = (s • x) ⊗ₜ[R] y := rfl
+
+variable {Z : Type v''} [AddCommGroup Z] [Module S Z]
+/--
+Let `R` be a commutative ring and `S` a ring.
+Given an `(S,R)`-bimodule `X` and a left `S`-module `Y`, the set of
+`S`-linear maps from `X` to `Y` has a left `R`-module structure given by:
+```
+l : X →ₗ[S] Y
+r : R
+x : X
+-------------
+(r • l) x = l (op r • x)
+```
+-/
+def Module.linearMapLeft : Module R (X →ₗ[S] Z) where
```

Comment:
This instance could be moved to the appropriate file (where LinearMap is defined), if not promoted to a global instance. 

On the other hand, `Module.tensorLeftOfMop` will be replaced by the correct version of `TensorProduct.leftModule`, and `Module.ofMop` will be gone after TensorProduct refactoring, so let's keep them here.

Reply:
This instance should be using `DomMulAct`, because it is acting on the domain of the linear map. With that change, it is *reasonably* safe as a global instance.

Reply:
Added some DomMulAct instances [here](https://github.com/leanprover-community/mathlib4/commit/0f9dd8841bc34771150aaf8eec70a7274c98aee7)